src/combinatorics/hall.lean

```diff
@@ -0,0 +1,382 @@
+  ∃ (f : α → β), function.injective f ∧ ∀ x, f x ∈ r x :=
+begin
+  tactic.unfreeze_local_instances,
+  revert α,
+  refine nat.strong_induction_on n (λ n' ih, _),
+  intros,
+  rcases n' with (_|_),
+  { exact hall_hard_inductive_zero r hn },
+  { apply hall_hard_inductive_step r n' hn hr,
+    introsI α' _ r' hα',
+    exact ih (fintype.card α') (nat.lt_succ_of_le hα') r' rfl, },
+end
+
+end hall_marriage_theorem
+
+/--
+This the version of Hall's Marriage Theorem in terms of indexed
+families of finite sets `s : ι → finset α`.  It states that there is
+set of distinct representatives if and only if every union of k of the
+sets has at least k elements.
+
+Recall that `A.bind s` is the union of all the sets `s i` for `i ∈ A`.
+-/
+theorem all_card_le_bind_card_iff_exists_injective
+  {ι α : Type*} [fintype ι] [decidable_eq α] (s : ι → finset α) :
+  (∀ (A : finset ι), A.card ≤ (A.bind s).card)
+  ↔ (∃ (f : ι → α), function.injective f ∧ ∀ x, f x ∈ s x) :=
+begin
+  split,
+  { exact hall_marriage_theorem.hall_hard_inductive s (fintype.card ι) rfl },
+  { rintro ⟨f, hf₁, hf₂⟩ A,
+    rw ←card_image_of_injective A hf₁,
+    apply card_le_of_subset,
+    intro b,
+    rw [mem_image, mem_bind],
+    rintros ⟨a, ha, rfl⟩,
+    exact ⟨a, ha, hf₂ a⟩, },
+end
+
+/-- Given a relation such that the image of every singleton set is finite, then the image of every
+finite set is finite. -/
+instance {α β : Type*} [decidable_eq β]
+  (r : α → β → Prop) [∀ (a : α), fintype (rel.image r {a})]
+  (A : finset α) : fintype (rel.image r A) :=
+begin
+  have h : rel.image r A = (A.bind (λ a, (rel.image r {a}).to_finset) : set β),
+  { ext, simp [rel.image], },
+  rw [h],
+  apply finset_coe.fintype,
+end
+
+/--
+This is a version of Hall's Marriage Theorem in terms of a relation
+between types `α` and `β` such that `α` is finite and the image of
+each `x : α` is finite (it suffices for `β` to be finite).  There is
+an injective function `α → β` respecting the relation iff for every
+`k` terms of `α` there are at least `k` terms of `β` related to them.
+
+If `[fintype β]`, then `[∀ (a : α), fintype (rel.image r {a})]` is automatically implied.
+-/
+theorem all_card_le_rel_image_card_iff_exists_injective
+  {α β : Type*} [fintype α] [decidable_eq β]
+  (r : α → β → Prop) [∀ (a : α), fintype (rel.image r {a})] :
+  (∀ (A : finset α), A.card ≤ fintype.card (rel.image r A))
+  ↔ (∃ (f : α → β), function.injective f ∧ ∀ x, r x (f x)) :=
+begin
+  let r' := λ a, (rel.image r {a}).to_finset,
+  have h : ∀ (A : finset α), fintype.card (rel.image r A) = (A.bind r').card,
+  { intro A,
+    rw ←set.to_finset_card,
+    apply congr_arg,
+    ext b,
+    simp [rel.image], },
+  have h' : ∀ (f : α → β) x, r x (f x) ↔ f x ∈ r' x,
+  { simp [rel.image], },
+  simp only [h, h'],
+  apply all_card_le_bind_card_iff_exists_injective,
+end
+
+/--
+This is a version of Hall's Marriage Theorem in terms of a relation
+between finite types.  It is like `hall_rel` but uses `finset.filter`
```

Comment:
What's `hall_rel`?

Reply:
That should be `fintype.all_card_le_rel_image_card_iff_exists_injective`